src/ship/GunManager.cpp

```diff
@@ -0,0 +1,519 @@
+}
+
+void GunManager::RemoveWeaponMount(StringName id)
+{
+	auto iter = m_mounts.find(id);
+	if (iter == m_mounts.end())
+		return;
+
+	if (IsWeaponMounted(id))
+		return;
+
+	m_mounts.erase(iter);
+}
+
+bool GunManager::MountWeapon(StringName hardpoint, const WeaponData &gunData)
+{
+	auto iter = m_mounts.find(hardpoint);
+	if (iter == m_mounts.end())
+		return false;
+
+	if (IsWeaponMounted(hardpoint))
+		return false;
+
+	WeaponState ws = {};
+
+	ws.mount = &iter->second;
+	ws.data = gunData;
+
+	// TODO: render weapon models on hardpoint
+	// if (!gunData.modelPath.empty()) {
+	// 	ws.model = Pi::FindModel(gunData.modelPath);
+	// }
+
+	// assign the new weapon to group zero
+	// TODO: make a new group for this weapon?
+	m_groups.front().weapons[m_weapons.size()] = true;
+	m_weapons.push_back(ws);
+
+	return true;
+}
+
+void GunManager::UnmountWeapon(StringName hardpoint)
+{
+	auto iter = std::find_if(m_weapons.begin(), m_weapons.end(), [&](const WeaponState &ws) {
+		return ws.mount->id == hardpoint;
+	});
+
+	if (iter != m_weapons.end()) {
+		uint32_t index = std::distance(m_weapons.begin(), iter);
+
+		// Update all group weapon assignments
+		for (auto &group : m_groups) {
+			RemoveGroupIndex(group.weapons, index);
+		}
+
+		// TODO: any additional cleanup needed for other systems that hook into GunManager, e.g. heat, targeting, etc.
+		m_weapons.erase(iter);
+	}
+}
+
+bool GunManager::IsWeaponMounted(StringName hardpoint) const
+{
+	return std::find_if(m_weapons.begin(), m_weapons.end(), [&](const WeaponState &ws) {
+		return ws.mount->id == hardpoint;
+	}) != m_weapons.end();
+}
+
+void GunManager::RemoveGroupIndex(WeaponIndexSet &group, uint32_t index)
+{
+	// Mask containing all of the indices below index
+	WeaponIndexSet keep = WeaponIndexSet().set() >> (NUM_WEAPON_INDICES - index);
+	// Mask containing all of the indices above index
+	WeaponIndexSet mask = WeaponIndexSet().set() << (index + 1);
+
+	// Decrement all contained indices in the set above the removed index down by one
+	group = (group & keep) | ((group & mask) >> 1);
+}
+
+uint32_t GunManager::GetWeaponIndexForHardpoint(StringName hardpoint) const
+{
+	auto iter = std::find_if(m_weapons.begin(), m_weapons.end(), [&](const WeaponState &ws) {
+		return ws.mount->id == hardpoint;
+	});
+
+	if (iter == m_weapons.end()) {
+		return UINT32_MAX;
+	}
+
+	return std::distance(m_weapons.begin(), iter);
+}
+
+void GunManager::SetupDefaultGroups()
+{
+	m_groups.clear();
+
+	GroupState group = {};
+
+	for (size_t i = 0; i < m_weapons.size(); i++) {
+		group.weapons[i] = true;
+	}
+
+	m_groups.push_back(group);
+}
+
+void GunManager::SetAllGroupsFiring(bool firing)
+{
+	for (size_t idx = 0; idx < m_groups.size(); idx++) {
+		SetGroupFiring(idx, firing);
+	}
+
+	m_isAnyFiring = firing;
+}
+
+void GunManager::AssignWeaponToGroup(uint32_t numWeapon, uint32_t group)
+{
+	if (numWeapon > m_weapons.size())
+		return;
+
+	if (group >= m_groups.size()) {
+		m_groups.resize(group + 1);
+	}
```

Comment:
Are we really this constrained by memory? Seems horribly inefficient to increase/decrease the weapon groups by 1, and, realistically, how many groups are we going to need? At most as many as there are weapons on the ship, or could a weapon belong to multiple groups?

I could imagine a player wanting weapons good against shields in one group, good against hull in another, or "all", for example.

In either case, it might make sense to pre-allocate the array. Conversely I guess this is not going to be called in a performance-critical codepath?

Reply:
Note that this is meant to handle the case where there is one extant group, and the caller decides that weapon 3 should be assigned to group 6.

> realistically, how many groups are we going to need?
> ...
> I could imagine a player wanting weapons good against shields in one group, good against hull in another, or "all", for example.

Group management is purposefully meant to be handled at a higher level than the GunManager. This API provides a low-level interface to assign weapons to numeric groups and control each group's target and firing state separately. There is a built-in convenience "shim" for operating on group 0 for ease of porting existing AI code, but the GunManager otherwise does not make any assumptions about how groups are going to be used or handled.

> At most as many as there are weapons on the ship, or could a weapon belong to multiple groups?

Each weapon can belong to at most one group - this is the main constraint on the system. I anticipate that higher-level code may want create a new group for a weapon on installation, but only for the player's ship.

> In either case, it might make sense to pre-allocate the array. Conversely I guess this is not going to be called in a performance-critical codepath?

Since the GunManager can't know how many weapons or groups will be used, it's not particularly helpful (or possible) to pre-allocate an array. Groups are also intended to be set once then forgotten; at most, one group reassignment is likely to happen per frame. Also note that the cost of `resize` is likely to be amortized as `std::vector` grows its internal storage based on a heuristic.

Reply:
I see; most ships will be AI so only have a single group, and the player will only setup groups rarely.

> Each weapon can belong to at most one group - this is the main constraint on the system. I anticipate that higher-level code may want create a new group for a weapon on installation, but only for the player's ship.

I can see this being a potential future limit; like I said, a player may want a "all" group for a full volley, for example. But possibly this can be expanded on in future if we decide we want it?

Reply:
> I can see this being a potential future limit; like I said, a player may want a "all" group for a full volley, for example. But possibly this can be expanded on in future if we decide we want it?

I should perhaps explain a bit more what I mean with "low level" - the player's "active weapons" state and trigger inputs are not meant to map 1:1 with the firing state of weapon groups in the GunManager. It's entirely possible to have a "volley fire" option with multiple weapon groups (internally it'd dispatch a fire command to all groups) or to have separate sets of weapons bound to different fire buttons, or have autonomous / auto-fire weapon groups with separate targeting.

The primary reason for the single-group constraint is so there's no need for a "priority system" to determine which group's inputs (firing, not firing) should be used to determine whether a weapon is firing or not.

TL;DR: Responsibility for dispatching input to individual fire-control groups is intentionally left to a higher-level layer than the GunManager, specifically so that we can do whatever we want.

Reply:
...okay.  Then I'm not entirely sure what the point of these low-level weapon groups are if the higher level code is expected to implement it's own grouping mechanism for the purposes of firing and button assignments etc. Sorry; colour me confused.

Reply:
These low-level groups are the building blocks of future higher-level "weapon behavior". If we (for example) wanted to have a system such that button A fires all long-range laser weapons, and button B fires all close-range ballistic weapons, that would be implemented by dispatching fire/stop-firing commands to separate groups. If we want a simple single-active-group system controlled by a single button, it'd be a thin shim over the existing functionality. If we decide that the player should manage each weapon separately, then we just assign weapons unique group indices and pretend like a weapon+group is the same "thing".

The system is implemented in this way at this time to defer any need to commit to a specific design for handling player fire-control to the future, once we've put some thought into the best way to handle it.

...my explanations are perhaps more confusing than they should be at this point - in short, it's best to consider it as something that will be revisited once we have a much more concrete understanding of our design requirements.

Reply:
I see; my assumption was that it was a given that weapon assignment to groups would be fully player-controlled. I didn't realise that wasn't decided yet.

Your last explanation clears this up and I (think I) understand the aim of this level implementation now.